airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/state/object_storage/ObjectStorageDestinationStateManager.kt

```diff
@@ -0,0 +1,103 @@
+import io.airbyte.cdk.load.state.DestinationStatePersister
+import io.airbyte.cdk.load.state.DestinationStatePersisterFactory
+import io.airbyte.cdk.load.util.serializeToString
+import io.airbyte.cdk.util.Jsons
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Secondary
+import jakarta.inject.Singleton
+
+class ObjectStorageDestinationState @JsonCreator constructor(generations: List<Generation>) :
+    DestinationState {
+    @JsonIgnore
+    private val generationMap: MutableMap<Long, MutableList<ObjectAndPart>> = mutableMapOf()
+
+    data class Generation(
+        @JsonProperty("generation_id") val generationId: Long,
+        @JsonProperty("objects") val objects: List<ObjectAndPart>
+    )
+
+    data class ObjectAndPart(
+        @JsonProperty("key") val key: String,
+        @JsonProperty("part_number") val partNumber: Long
+    )
+
+    @get:JsonProperty("generations")
+    val generations: List<Generation>
+        get() = generationMap.map { Generation(it.key, it.value) }
+
+    init {
+        generations.forEach { generationMap[it.generationId] = it.objects.toMutableList() }
+    }
+
+    // Methods to modify the map directly
+    suspend fun addObject(generationId: Long, key: String, partNumber: Long) {
+        val objects = generationMap.getOrPut(generationId) { mutableListOf() }
+        objects.add(ObjectAndPart(key, partNumber))
+    }
+
+    suspend fun removeObject(generationId: Long, key: String) {
+        generationMap[generationId]?.removeIf { it.key == key }
+    }
+
+    // Accessor for loaded data in Map format
+    fun getGenerationMap(): Map<Long, List<ObjectAndPart>> = generationMap
+}
+
+class ObjectStorageStagingPersister(
+    private val client: ObjectStorageClient<*>,
+    private val pathFactory: PathFactory
+) : DestinationStatePersister<ObjectStorageDestinationState> {
+    private val log = KotlinLogging.logger {}
+
+    companion object {
+        const val STATE_FILENAME = "__airbyte_state.json"
+    }
+
+    private fun keyFor(stream: DestinationStream): String =
+        pathFactory.getStagingDirectory(stream).resolve(STATE_FILENAME).toString()
+
+    override suspend fun load(stream: DestinationStream): ObjectStorageDestinationState {
+        val key = keyFor(stream)
+        try {
+            log.info { "Loading destination state from $key" }
+            return client.get(key) { inputStream ->
+                Jsons.readTree(inputStream).let {
+                    Jsons.treeToValue(it, ObjectStorageDestinationState::class.java)
+                }
+            }
+        } catch (e: Exception) {
+            log.info { "No destination state found at $key" }
+            return ObjectStorageDestinationState(emptyList())
+        }
+    }
+
+    override suspend fun persist(stream: DestinationStream, state: ObjectStorageDestinationState) {
+        val payload = Jsons.valueToTree<ObjectNode>(state).serializeToString()
```

Comment:
nit: I think you can just `Jsons.writeValueAsString(state)`?

Reply:
```
fun Any.serializeToJsonBytes(): ByteArray {
    return Jsons.writeValueAsBytes(this)
}
```
and
```
    override suspend fun persist(stream: DestinationStream, state: ObjectStorageDestinationState) {
        client.put(keyFor(stream), state.serializeToJsonBytes())
    }
```
😏